driver-core/src/main/com/mongodb/operation/ChangeStreamBatchCursor.java

```diff
@@ -125,13 +153,22 @@ public void remove() {
                 if (!rawDocument.containsKey("_id")) {
                     throw new MongoChangeStreamException("Cannot provide resume functionality when the resume token is missing.");
                 }
-                resumeToken = rawDocument.getDocument("_id");
                 results.add(rawDocument.decode(changeStreamOperation.getDecoder()));
             }
+            cacheResumeToken(rawDocuments);
         }
         return results;
     }
 
+    private void cacheResumeToken(final List<RawBsonDocument> rawDocuments) {
+        RawBsonDocument lastDocument = rawDocuments.get(rawDocuments.size() - 1);
+        if (lastDocument.containsKey("postBatchResumeToken")) {
```

Comment:
Will this ever be the case? If not you can probably revert back to setting the resumeToken in convertResults.

Reply:
The spec does state that the last document may contain the `postBatchResumeToken`:
```
* When returning a document to the user:
    * If it's the last document in the batch and a postBatchResumeToken is included, cache it.
    * Else, cache the _id of the document.
```

Reply:
That's saying that you should ignore the _id of the last document in the batch if there is a postBatchResumeToken.  Regardless, as discussed this logic will have to move to MongoChangeStreamCursorImpl.

Reply:
Thanks for the clarification. That's clever, so if you filtered the results presumably the `postBatchResumeToken` would be ahead to the last `_id` nice.

Reply:
If the QueryBatchCursor found a postBatchResumeToken, it will be cached here.